### PR TITLE
Add TCP logging output

### PR DESCRIPTION
**Pull request by tabboud**
## Before this PR
No support to publish logs to a network log target.

## After this PR

==COMMIT_MSG==
Add support to write JSON encoded log lines to a TCP socket and expose service-discovery runtime config.

Services can now declare an "sls-log-tcp-json-receiver" service-discovery block, where logs will be multiplexed between the TCP receiver and the default log output destination. The log lines will be wrapped in an envelope that contains metadata about the source origin, including hostname, product, product version, etc. The TCP receiver is expected to use TLS authentication so valid security material must also be configured under the service-discovery block.
==COMMIT_MSG==

## Notes about the architecture
- All but the `service-discovery` config is internal for the TCP writer, so we will be free to make any changes without worrying about breaking consumers.
- Rather than go for a full blown TCP connection pool, I decided to implement a simple TCP connection provider that will round robin the set of URIs, establish a connection and return the connection to the writer. It is up to the writer to handle the lifecycle of the `net.Conn`. `net.Conn` can be used concurrently, so it is currently cached on the writer and a new one is retrieved if a connection is closed.
- All loggers share a single `TCPWriter`, which is in line with the Java implementation. We could make it so each logger has it's own `TCPWriter` especially if there is a lot of contention. This would require significantly more changes to how we instantiate our loggers today and would increase the overall number of outgoing connections.
- The envelope metadata is static, but we currently `json.Marshal` this along with the payload on every write, which is not very efficient. However without doing some manual JSON manipulation (i.e. serialize the envelope then remove the last "}" and jam in the `,"payload":payload}`), there is not an easy way to pre-serialize the metadata.
- The `TCPWriter` implements the `io.Writer` interface which lets it be used for the wlog output. Since witchcraft-go-server uses wlog-zap internally we are safe to assume all log output is JSON. However if we ever change that then this implied API will be broken. Also based on the zap docs, each call to `Write()` should contain an entire log payload, which deviates slightly from the `io.Writer` interface. See `Entry` here: https://pkg.go.dev/go.uber.org/zap@v1.16.0/zapcore#Core.

## Possible downsides?

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-server/251)

### CODE REVIEW

File: config/runtime.go
Comment (tabboud) [inline config/runtime.go]: Given this is a public facing addition, I am happy to make this a pre-commit to differentiate between the internal TCP writer things. Although this shows how it would be used.

File: witchcraft/internal/tcpjson/conn_provider.go
Comment (nmiyake) [inline witchcraft/internal/tcpjson/conn_provider.go]: nit: "users" -> "caller's"

Comment (nmiyake) [reply witchcraft/internal/tcpjson/conn_provider.go]: Is it safe for any given caller to close the connection, or will doing so possibly close the connection used by all callers?

File: witchcraft/internal/tcpjson/conn_provider.go
Comment (nmiyake) [inline witchcraft/internal/tcpjson/conn_provider.go]: Since `ConnProvider` is the interface that callers interact with, this struct can probably be private

File: witchcraft/internal/tcpjson/conn_provider.go
Comment (nmiyake) [inline witchcraft/internal/tcpjson/conn_provider.go]: Return value can be `ConnProvider`

File: witchcraft/internal/tcpjson/conn_provider.go
Comment (nmiyake) [inline witchcraft/internal/tcpjson/conn_provider.go]: Include the string of the URI that failed to parse in the output so that it's easier to debug/determine what went wrong

File: witchcraft/internal/tcpjson/conn_provider.go
Comment (nmiyake) [inline witchcraft/internal/tcpjson/conn_provider.go]: Comment would be helpful -- note that this is the index of the next host that will be handed out

File: witchcraft/internal/tcpjson/conn_provider.go
Comment (nmiyake) [inline witchcraft/internal/tcpjson/conn_provider.go]: Current logic works, but is a bit confusing to follow -- specifically, it's a bit strange to me that the idx being stored is the "next" one, but before the mod operation is applied, meaning that the value in the struct as stored may actually be out of bounds until the mod operation is applied.

Think it would be cleaner to have the `s.hostIdx` be the index of the next host that will be handed out, and then apply the mod operation before updating.

If the logic is written this way to support the case where the length of `s.hosts` might change between calls then maybe the current approach makes more sense, but doubt that is the case given that there isn't any code to guarantee the safety around that slice between accesses (and if this is the case, then I think it would be helpful to have a comment to explain so).

File: witchcraft/internal/tcpjson/conn_provider_test.go
Comment (nmiyake) [inline witchcraft/internal/tcpjson/conn_provider_test.go:76]: Can you add a brief comment to explain what this is testing?

File: witchcraft/internal/tcpjson/envelope.go
Comment (nmiyake) [inline witchcraft/internal/tcpjson/envelope.go]: nit: should be `SLSEnvelopeV1`.

Is this specification taken from somewhere? If so, might be helpful to include link.

File: witchcraft/witchcraft.go
Comment (nmiyake) [inline witchcraft/witchcraft.go]: Can we add message/context to error that says that security material could not be created for the TCP JSON receiver?

File: witchcraft/witchcraft.go
Comment (nmiyake) [inline witchcraft/witchcraft.go:638]: Do we know that running this again is safe? Would it make more sense to do the TCP writer logic first and then pass `tcpWriter` to the place where `initLoggers` is currently called?

File: witchcraft/witchcraft.go
Comment (nmiyake) [inline witchcraft/witchcraft.go:607]: Is there any logic that logs on the context loggers in `s.initRuntimeConfig`? That would be impacted by the removal here.

File: witchcraft/witchcraft.go
Comment (nmiyake) [reply witchcraft/witchcraft.go:638]: Ah OK I see, this requires the runtime config to be loaded first. Then would just want to double-check the contract for the init loggers call.

File: witchcraft/internal/tcpjson/tcp_logger.go
Comment (nmiyake) [inline witchcraft/internal/tcpjson/tcp_logger.go]: Should we use `safejson.Marshal` here? I seem to recall that the default adds a newline and doesn't handle certain encodings etc. (which was our motivation for having `safejson`)

File: witchcraft/internal/tcpjson/tcp_logger.go
Comment (tabboud) [inline witchcraft/internal/tcpjson/tcp_logger.go]: I am starting to realize that this breaks the contract for an `io.Writer` since we end up writing more than `len(logPayload)` bytes to the underlying connection.

**Review by nmiyake — CHANGES_REQUESTED** [review]
Thanks for submitting this -- was very clear to follow, especially with the overall contextual comment. I think the overall approach looks good. High-level feedback:

* Yes, would appreciate separate PR that adds the configuration first/separately
* Wanted to clarify a bit about the last bullet point in your overview. Your point is basically that since the TCP writer wraps the loggers, if the wrapped loggers don't emit JSON then the TCP writer will wrap a payload that's not actually JSON -- is that correct? If so, I think that's acceptable -- the contract can just be stipulated that the TCP writer's payload will be the output of the underlying loggers, whether that be JSON, text, or anything else. Since the envelope itself is fully marshalled as JSON, I think the output will still be valid JSON even if the payload isn't, which I think is the only guarantee that's important.

I am actually a bit concerned about the use of `json.Marshal` -- I think that's known to be a pretty inefficient implementation in terms of both space and time, and since logging is such a hot path I think it could be significant.

Could we write benchmarks to test this? The main other approaches that come to mind are pre-creating the JSON wrapper and printing the provided bytes into it (but then downside is that if provided bytes are not valid JSON then output will not be either), or even using a zerolog instance here that constructs the output that way.

File: witchcraft/internal/tcpjson/tcp_logger.go
Comment (nmiyake) [reply witchcraft/internal/tcpjson/tcp_logger.go]: I think this should be OK from a contract standpoint -- the API reports the number of input bytes written/processed, so the amount that we write to the underlying connection doesn't really matter.

However, it is true that the return value should reflect that (so in this example, if we write all the input, then it needs to return `len(logPayload)` rather than the number of byte that were written to the connection)

Comment (nmiyake) [reply witchcraft/internal/tcpjson/tcp_logger.go]: For reference, https://stackoverflow.com/questions/32571504/what-number-of-bytes-written-should-my-io-writer-return-when-it-intentionally#comment52998337_32571504

Comment (tabboud) [reply witchcraft/internal/tcpjson/tcp_logger.go]: Thanks for clarifying this. I was getting tripped up thinking all io.Writer's had to return the total bytes actually written rather than the bytes processed of the original input.

File: witchcraft/internal/tcpjson/conn_provider.go
Comment (tabboud) [reply witchcraft/internal/tcpjson/conn_provider.go]: Doing so will close the connection for all callers. If the connection is closed between the time of retrieving and then writing then a "use of closed connection" error will be returned. Added this to the doc comment as well.

In our case the TCPWriter will catch this error and will drop the log line, which may not be ideal. In order to get around this, we would have to not share a single `net.Conn` across all writers, but instead have each writer take a lock on each `net.Conn`, so we would lose 1 log line at the max.

File: witchcraft/internal/tcpjson/conn_provider.go
Comment (tabboud) [reply witchcraft/internal/tcpjson/conn_provider.go]: I usually like to return concrete types in situations where an implementation could satisfy multiple interfaces rather than defining a single large interface that groups everything. In this case there is only 1 method, so I am fine to do make it private and return the interface.

File: witchcraft/internal/tcpjson/conn_provider.go
Comment (tabboud) [reply witchcraft/internal/tcpjson/conn_provider.go]: I see how that can be confusing and think your suggestion makes a lot of sense. Given we're currently expecting `s.hosts` to be static, I am going to update with your suggestion and always store the next index to use and perform the mod before storing. This way the index we use is just whatever the current value is.

File: witchcraft/internal/tcpjson/envelope.go
Comment (tabboud) [reply witchcraft/internal/tcpjson/envelope.go]: This is mirrored from the internal sls-spec, which has not been migrated to witchcraft-api yet.
I updated this to `LogEnvelopeV1` instead of SLS since we're re-defining it internally.

File: witchcraft/witchcraft.go
Comment (tabboud) [reply witchcraft/witchcraft.go:638]: Yep that is the correct observation. The `initLoggers` call only operates over the logger objects, so it's safe to re-run this again given we shouldn't have to worry about concurrent access or anything getting clobbered. However, you did point out a bug that I expanded on below.

File: witchcraft/witchcraft.go
Comment (tabboud) [reply witchcraft/witchcraft.go:607]: Good catch! The `s.initRuntimeConfig` does need the loggers on the context, so I moved it back to where it was. I also looked into what `s.withLoggers` is doing and it appears safe to re-run after we re-initialize the loggers for the TCP writer, so I added an additional call below which will overwrite the existing logger on the context.

File: witchcraft/internal/tcpjson/tcp_logger.go
Comment (tabboud) [reply witchcraft/internal/tcpjson/tcp_logger.go]: unfortunately not because `safejson.Marshal` uses `json.Encoder` under the hood and then strips off the `\n` char that we need to keep. I have since replaced this entirely as well.

**Comment by tabboud** [comment]
@nmiyake I believe I have addressed all of the comments and this should be ready for another round of review.

To address your 2 bullet points
- I went ahead and opened up #252 which adds in the `service-discovery` config as a pre-commit. I will remove it from here once we merge that later today.
- Yes this is correct. We expect that the payload is JSON, but if it is not we will still wrap it in the envelope.

I added in a benchmark for the serializers to see what the result would be between json.Marshal, zerolog, or a manual formatting. Here are the results:

```
pkg: github.com/palantir/witchcraft-go-server/v2/witchcraft/internal/tcpjson
BenchmarkEnvelopeSerializer
BenchmarkEnvelopeSerializer/zerolog
BenchmarkEnvelopeSerializer/zerolog-16         	 2279970	       530 ns/op	    1505 B/op	       5 allocs/op
BenchmarkEnvelopeSerializer/JSON-Encoder
BenchmarkEnvelopeSerializer/JSON-Encoder-16    	  778161	      1609 ns/op	     688 B/op	       3 allocs/op
BenchmarkEnvelopeSerializer/JSON-Marshaler
BenchmarkEnvelopeSerializer/JSON-Marshaler-16  	  698350	      1523 ns/op	     640 B/op	       2 allocs/op
BenchmarkEnvelopeSerializer/manual
BenchmarkEnvelopeSerializer/manual-16          	 6416482	       175 ns/op	     768 B/op	       1 allocs/op
```

Overall the manual injection is approximately memory and CPU efficient than all other approaches, with the side affect of how hacky it is. The zerolog serializer (as it stands right now) is much more CPU efficient, but does allocate a bit more than regular json marshaling, which is interesting. In the end I opted for zerolog, since it seems like a nice middle ground between code complexity and performance. I currently use a bytes buffer for each serialization, where we could use a buffer pool to save some cost, but deferred that for now.

Also, another interesting thing we could do is write directly to the `net.Conn`, since it supports concurrent writes by default, rather than use a byte buffer. For each zerolog invocation we update the output, which can be any io.Writer, to be the local buffer but we could pass it the `net.Conn` and let it handle the writing. The challenge here becomes how to handle the transient errors and close the `net.Conn` appropriately.

File: witchcraft/internal/tcpjson/conn_provider.go
Comment (nmiyake) [inline witchcraft/internal/tcpjson/conn_provider.go]: typo: should be "caller's"

File: witchcraft/internal/tcpjson/conn_provider.go
Comment (nmiyake) [inline witchcraft/internal/tcpjson/conn_provider.go:55]: Add documentation for function that describes that:
* Returns a new `ConnProvider`
* Provided set of uris must have at least one element and must be able to be parsed as URIs
* If config is `nil`, then default is used

File: witchcraft/internal/tcpjson/tcp_logger.go
Comment (nmiyake) [inline witchcraft/internal/tcpjson/tcp_logger.go:49]: Comment please

File: witchcraft/internal/tcpjson/tcp_logger.go
Comment (nmiyake) [inline witchcraft/internal/tcpjson/tcp_logger.go:66]: Think comment would be helpful for this as well: describe that input is wrapped, written to connection, behavior if there is a connection error etc.

File: witchcraft/internal/tcpjson/tcp_logger.go
Comment (nmiyake) [inline witchcraft/internal/tcpjson/tcp_logger.go]: nit: grammar

```
// create a new top-level logger that writes to ioutil.Discard since each
// serialization will write to its own local buffer instead of a single writer
```

File: witchcraft/loggers.go
Comment (nmiyake) [inline witchcraft/loggers.go:45]: Update function documentation to describe `tcpWriter` param

**Review by nmiyake — CHANGES_REQUESTED** [review]
Thanks -- just some minor feedback in terms of content, and as discussed will wait for config PR to merge before this one.

File: witchcraft/internal/tcpjson/tcp_logger.go
Comment (tabboud) [reply witchcraft/internal/tcpjson/tcp_logger.go:49]: I had a comment on the exported type, so moved that one down here instead.

File: witchcraft/internal/tcpjson/tcp_logger.go
Comment (tabboud) [reply witchcraft/internal/tcpjson/tcp_logger.go]: 🤦‍♂️ fixed

File: witchcraft/internal/tcpjson/envelope.go
Comment (nmiyake) [inline witchcraft/internal/tcpjson/envelope.go]: I think the `json":,inline"` doesn't actually do anything, so might as well remove

File: witchcraft/internal/tcpjson/tcp_logger.go
Comment (nmiyake) [inline witchcraft/internal/tcpjson/tcp_logger.go]: The https://golang.org/pkg/bytes/#TrimSuffix function basically does this -- can do `bytes.TrimSuffix(b, []byte("\n"))` in place of calling this function.

**Review by nmiyake — CHANGES_REQUESTED** [review]
Minor feedback

File: witchcraft/internal/tcpjson/envelope.go
Comment (tabboud) [reply witchcraft/internal/tcpjson/envelope.go]: I saw the issue about that actually which is what made me embed the type, just forgot to remove the tag. thanks!

File: witchcraft/internal/tcpjson/tcp_logger.go
Comment (tabboud) [reply witchcraft/internal/tcpjson/tcp_logger.go]: I intentionally wrote this to avoid allocating yet another `[]byte`, but in the grand scheme of things feels like an over optimization. Went ahead and changed it to the stdlib function.

Comment (nmiyake) [reply witchcraft/internal/tcpjson/tcp_logger.go]: By allocation, do you mean the `[]byte("\n")` line (since I don't believe the implementation of the function performs any allocations)? If we think that's an issue, then fine with going with this function, although would explain that in a comment. If the allocation is the only concern we could also have a package-level `var newlineBytes = []byte("\n")` that is allocated once and used in the trim operation

Comment (tabboud) [reply witchcraft/internal/tcpjson/tcp_logger.go]: Yep I was referring to the `[]byte("\n")`. I ran a quick benchmark to check my hypothesis and it turns out inlining the `[]byte("\n")` does not actually allocate, so no need for the local var.

**Comment by bmoylan** [comment]
___
*[witchcraft/internal/tcpjson/conn_provider.go, line 81 at r9](https://reviewable.io/reviews/palantir/witchcraft-go-server/251#-MUVB4Zl-RqBeEMmIWof:-MUVB4Zl-RqBeEMmIWog:b-bziv8p) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/5eb72d299b777e5f42e481f018b65527580fe0d8/witchcraft/internal/tcpjson/conn_provider.go#L81)):*
> ```Go
> 	atomic.CompareAndSwapInt32(&s.nextHostIdx, hostIdx, int32(nextHostIdx))
> 
> 	tlsConn, err := tls.Dial("tcp", s.hosts[hostIdx], s.tlsConfig)
> ```

we probably want to set some timeout on the Dial so it can't just hang forever. This requires constructing a dialer struct like
```go
tlsDialer := &tls.Dialer{NetDialer: &net.Dialer{Timeout: 10*time.Second}, Config: tlsConfig}
```

**Review by tabboud — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-server/251)* status: 0 of 49 files reviewed, 18 unresolved discussions (waiting on @bmoylan, @nmiyake, and @tabboud)

___
*[witchcraft/internal/tcpjson/conn_provider.go, line 81 at r9](https://reviewable.io/reviews/palantir/witchcraft-go-server/251#-MUVB4Zl-RqBeEMmIWof:-MUoIkEK8RL7VoNcB7h-:b6anvs8) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/5eb72d299b777e5f42e481f018b65527580fe0d8/witchcraft/internal/tcpjson/conn_provider.go#L81)):*
<details><summary><i>Previously, bmoylan (Brad Moylan) wrote…</i></summary><blockquote>

we probably want to set some timeout on the Dial so it can't just hang forever. This requires constructing a dialer struct like
```go
tlsDialer := &tls.Dialer{NetDialer: &net.Dialer{Timeout: 10*time.Second}, Config: tlsConfig}
```
</blockquote></details>

Updated to use `tls.DialWithDialer` since `tls.Dialer` was not added until go1.15 (https://golang.org/pkg/crypto/tls/#Dialer). Once we update and vet the circle images to bump from go1.14 as the min, then we can update this as well

**Review by nmiyake — APPROVED** [review]

**Comment by tabboud** [comment]
@bmoylan Did you want to take another pass before merging? 
I am going to make a follow up PR with a health check that monitors the active connection for the TCP writer (i.e. checking if tcpWriter.conn != nil).

**Review by bmoylan — APPROVED** [review]
> Reviewed 2 of 10 files at r1, 41 of 43 files at r3, 2 of 6 files at r4, 1 of 2 files at r6, 1 of 2 files at r7, 3 of 3 files at r8, 1 of 1 files at r10.
> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-server/251)* status: all files reviewed, 17 unresolved discussions (waiting on @nmiyake and @tabboud)

**Comment by bmoylan** [comment]
Let's do it!

**Pull request merged**
